### DISCUSSION TITLE
Why shallow clone is expensive to perform?

### DISCUSSION THREAD

**Discussion by xxyzz** [Homebrew/discussions#225] [Everyday usage] [answered]
I find [this commit](https://github.com/Homebrew/brew/commit/10e73164f0f00c0451cd7c8ca289960d1326f7a0) that requires homebrew-core and  homebrew-cask taps to be unshallow cloned but I don't why it's expensive.

**Comment by jonchang** [comment, marked as answer]
See https://github.com/CocoaPods/CocoaPods/issues/4989#issuecomment-193772935 and https://public-inbox.org/git/20160307221539.GA24034@sigill.intra.peff.net/T/. Every `brew update` that fetches into a shallow clone has a big performance and bandwidth penalty.

**Reply by xxyzz** [reply]
Thanks! mhagger's following [comment](https://github.com/CocoaPods/CocoaPods/issues/4989#issuecomment-193786578) is also quite helpful. What if the subsequent fetches all use `--depth=1`, will it have performance issue?

**Reply by gromgit** [reply]
From the `git-fetch` man page:
> `--depth=<depth>`
>
> Limit fetching to the specified number of commits from the tip of each remote branch history. If fetching to a _shallow_ repository created by `git clone` with `--depth=<depth>` option (see `git-clone(1)`), deepen or shorten the history to the specified number of commits. Tags for the deepened commits are not fetched.

So if you `git fetch --depth=1` on a shallow core repo (cloned `--depth=1`), you haven't improved the situation at all.

**Reply by xxyzz** [reply]
So the best way to update a shallow cloned repo is delete the folder then shallow clone again?

**Reply by gromgit** [reply]
That _might_ be the most _efficient_ way to update a shallow clone, but in the general case, that would lose any changes you'd made in the working directory, so "best" is a stretch at best.

And, of course, if you pull that sort of stunt on the Homebrew core repo, you'd be inviting GitHub to drop the rate-limit hammer on all users...including yourself.

**Reply by SMillerDev** [reply]
The best way to update a shallow clone of homebrew is to unshallow it and use it that way.

**Reply by chrisfinazzo** [reply]
Huh.

Although I usually run `update` - and also fetch the upstream -  into my copy of Core and/or Cask a couple times a day, I first encountered this late yesterday.

In that case, would it be unwise to alias `brew update`  as `git -C "/usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask" fetch --unshallow && brew update` or is there another way to get around this?

**Reply by SMillerDev** [reply]
You just need to unshallow it once.

**Reply by duhd1993** [reply]
``/usr/local/Homebrew/Library/Taps/`` is already almost 1GB on my mac. Although disk is not that expensive nowdays (I guess for Mac it's still expensive), this size will only increase with time. Do you have any plan for this? Most of the disk space used is actually not useful for normal users.

**Reply by carlocab** [reply]
Using an unshallow clone is the price for GitHub not imposing rate-limits on access to Homebrew repositories. I think rate-limits would create a terrible experience for normal users, so I think it absolutely is useful.

If you want to cut down on the storage space that your taps use, you can try `git gc --aggressive`. (Caveat: please read up on what this does before trying it yourself.)

**Comment by ghost** [comment]
Theoretically, for a pure linear commit history (without merges), following fetch from a shallow clone shouldn't be expensive.

Could something be improved for git itself?

**Reply by SMillerDev** [reply]
If you can find a way to optimize this in Git i'm sure GitHub engineers would be happy with it.

**Reply by carlocab** [reply]
I guess the difficulty is that `git` has no guarantee that a repo has a pure linear commit history. It also has no way of guaranteeing that a repo with such a history will stay that way.

**Comment by yichengchen** [comment]
https://blog.cocoapods.org/CocoaPods-1.7.2/ CocoaPods already support http CDN for years and greatly improve the user experience. 
I guess it's the right way brew should follow.

**Reply by SMillerDev** [reply]
That's just over a year since the blog post. But if you have an idea for an implementation you can make a pull request.
